### PR TITLE
dialyzer: Introduce nominal types

### PR DESCRIPTION
**Pull request by lucioleKi**
Co-authored-by: John Högberg <john@erlang.org>

This PR implements [EEP-69: Nominal Types](https://www.erlang.org/eeps/eep-0069)
in Dialyzer. As a side effect, nominal types can encode opaque types. We
changed all opaque-handling logic and improved opaque warnings in
Dialyzer.

All existing Erlang type systems are structural: two types are seen as
equivalent if their structures are the same. Type comparison are based
on the structures of the types, not on how the user explicitly defines
them. For example, in the following example, `meter()` and `foot()` are
equivalent. The two types can be used interchangeably. Neither of them
differ from the basic type `integer()`.

    -type meter() :: integer().
    -type foot() :: integer().

Nominal typing is an alternative type system, where two types are
equivalent if and only if they are declared with the same type name.
The EEP proposes one new syntax `-nominal` for declaring nominal types.
Under nominal typing, `meter()` and `foot()` are no longer compatible.
Whenever a function expects type `meter()`, passing in type `foot()`
would result in a Dialyzer error.

    -nominal meter() :: integer().
    -nominal foot() :: integer().

More nominal type-checking rules can be found in the EEP. It is worth
noting that most work for adding nominal types and type-checking
is in `erl_types.erl`. The rest are changes that removed the previous
opaque type-checking, and added an improved version of it using nominal
type-checking with reworked warnings.

Backwards compatibility for opaque type-checking is **not** preserved
by this PR. Previous opaque warnings can appear with slightly different
wordings. A new kind of opaque warning `opaque_union` is added, together
with a Dialyzer option `no_opaque_union` to turn this kind of warnings
off.

This PR is likely not bug-free, but not having it in master is blocking
other changes to Dialyzer. Once nominal type-checking and the new opaque
type-checking are applied to more codebases, we will fix the bugs as
they appear.

### CODE REVIEW

**Comment by bjorng** [comment]
Is it really necessary to update the primary bootstrap as part of this pull request?

File: lib/dialyzer/src/erl_types.erl [hunk 1]
```diff
@@ -288,15 +278,29 @@
...
+
+%% Special type used to mark infinite loops: functions are assumed to return
+%% a supertype of ?unit rather than ?none during analysis, letting us
+%% distingish between functions that intentionally never return (like server
+%% loops) and functions that never return because of a crash.
 -define(unit, unit).
+
+%% Special type used to mark opaque nominals during opacity violation checking.
+%%
+%% FIXME: Rename this to ?opaque once the old opaques have been chucked out.
```
Comment (TD5) [inline lib/dialyzer/src/erl_types.erl:295]: Looks like this this FIXME is done?

**Comment by jhogberg** [comment]
> Is it really necessary to update the primary bootstrap as part of this pull request?

Yes, `erl_lint` actually checks that `-dialyzer( ... )` directives don't contain any unknown options, which broke the build when we added `-dialyzer(no_opaque_union, ...)` directives to suppress certain warnings in OTP itself (which would otherwise have broken CI, because `dialyzer` has to run clean).

File: lib/dialyzer/src/erl_types.erl [hunk 2]
```diff
@@ -419,347 +418,86 @@ t_is_none(_) -> false.
...
+  %% blinded infimum of `Given` and `Required`.
+  Direction = case t_is_impossible(t_inf(Given, Required)) of
+                true -> ?any;
+                false -> ?opaque
+              end,
+
+  RequiredBlind = oc_mark(Required, Direction, Module),
+  GivenBlind = oc_mark(Given, Direction, Module),
+
+  %% If the `Required` type does not change when blinded, we know that the call
```
Comment (TD5) [inline lib/dialyzer/src/erl_types.erl:453]: How do we expect this logic to handle complex types which Dialyzer "simplifies", in terms of genuine issues not being raised due to type simplification vs. issues falsely raised due type simplification?

**Comment by TD5** [comment]
This is the most exciting change I've seen to Dialyzer's semantics... ever? Very cool stuff! 😎 I am definitely keen to try it out for myself.

What is your expectation regarding performance vs. previous logic for opaques? Should we expect an appreciable speed up / slow down of a Dialyzer analysis as a result of this change?

**Comment by TD5** [comment]
With the introduction of nominal types, perhaps we should revisit any existing docs or other material which still claim that Dialyzer only reports "genuine" type errors in order to clarify that a Dialyzer complaint doesn't necessarily mean code will crash for any input at runtime, but rather there's a discrepancy between definition and use (e.g. due to record definitions, specs & nominal types). I am not sure whether any "official" docs still say something to that effect, but my impression is that it is a notion still circulating in the community.

File: lib/dialyzer/src/erl_types.erl (hunk 2 above)
Comment (jhogberg) [reply lib/dialyzer/src/erl_types.erl:453]: We'll only see such simplifications where we have unions of opaques and non-opaques, which is iffy enough that we're adding a  warning for that in this PR, so I don't expect this to become a huge problem.

**Comment by jhogberg** [comment]
> What is your expectation regarding performance vs. previous logic for opaques? Should we expect an appreciable speed up / slow down of a Dialyzer analysis as a result of this change?

For most code there shouldn't be much of a difference, but the insane corner cases with quartic runtime are gone. I'd expect https://github.com/erlang/otp/issues/7835 to be resolved with this PR.

> With the introduction of nominal types, perhaps we should revisit any existing docs or other material which still claim that Dialyzer only reports "genuine" type errors in order to clarify that a Dialyzer complaint doesn't necessarily mean code will crash for any input at runtime, but rather there's a discrepancy between definition and use (e.g. due to record definitions, specs & nominal types). I am not sure whether any "official" docs still say something to that effect, but my impression is that it is a notion still circulating in the community.

That was never true for opaques to begin with, but yes, we should correct any documentation that makes those claims.

**Review by TD5 — APPROVED** [review]
Not that I am a maintainer, but I have spent some time poking around inside Dialyzer, and _I think this change makes sense_, certainly given the linked EEP, and it is surprisingly uninvasive, all things considered. 

I buy into the idea of landing this PR and fixing forward, since keeping this sort of change out of master just means stacking up more potential conflicts.

File: lib/dialyzer/src/erl_types.erl (hunk 2 above)
Comment (lucioleKi) [reply lib/dialyzer/src/erl_types.erl:453]: An opaque with inner type `any` is still an opaque. `erl_types:t_opacity_conflict/3` will detect the opaque violation regardless of whether the opaque was simplified by `t_limit/2`. During testing, we had false positives due to our incorrect treatment for nominals in `t_limit/2`. We'll see if new ones pop up.

Comment (TD5) [reply lib/dialyzer/src/erl_types.erl:453]: Ah, yes, that's exactly what I was getting at! Sounds like you've considered those cases I was wondering about :)

**Review by bjorng — APPROVED** [review]

File: lib/dialyzer/src/erl_types.erl [hunk 3]
```diff
@@ -2758,81 +2468,151 @@ t_inf([]) -> ?none.
...
-t_inf(?nil, ?nonempty_list(_, _), _Opaques) ->
+  t_map(Pairs,
+        t_inf_aux(ADefK, BDefK),
+        t_inf_aux(ADefV, BDefV));
+%% Intersection of 1 or more nominal types
+t_inf_aux(?nominal(Same, LHS_S), ?nominal(Same, RHS_S)) ->
+  t_nominal(Same, t_inf_aux(LHS_S, RHS_S));
+t_inf_aux(?nominal(LHS_Name, ?nominal(LHS_InnerName, _)=LHS_Inner),
+          ?nominal(RHS_Name, ?nominal(RHS_InnerName, _)=RHS_Inner)) ->
+  %% FIXME: Explain this in more detail;
```
[the remark this answers is not in the document]
Comment (jhogberg) [reply lib/dialyzer/src/erl_types.erl:2540]: Oops, forgot about that one. This should do it:

```
  %% As the names of these nominals differ, they can only intersect if LHS is
  %% a nominal subtype of RHS_Inner or if RHS is a nominal subtype of LHS
  %% inner, for example:
  %%
  %% t_nominal(alpha, t_nominal(beta, t_nominal(gamma, any()))) = LHS
  %%                  t_nominal(beta, t_nominal(gamma, t_atom())) = RHS
  %%   =>
  %% t_nominal(alpha, t_nominal(beta, t_nominal(gamma, t_atom()))) = Res
  %%
  %% Note that nested nominals only intersect with nominals that share the
  %% same nesting: in a sense, you can say that the effective name of a nominal
  %% is the sum of its nesting. Thus, the following do not intersect despite
  %% being `alpha`s that are subtype of `gamma`s:
  %%
  %% t_nominal(alpha, t_nominal(beta, t_nominal(gamma, any()))) = LHS
  %% t_nominal(alpha, t_nominal(gamma, t_atom())) = RHS
  %%
  %% These rules are described in "Nominal Types for Erlang" by Huang et al,
  %% https://doi.org/10.1145/3677995.3678191
```

File: lib/dialyzer/src/erl_types.erl [hunk 4]
```diff
@@ -3735,74 +3432,54 @@ t_is_equal(_, _) -> false.
...
-             case t_contains_opaque(T2) of
-               true  -> t_unopaque(T2);
-               false -> T2
-             end).
+  Inf = t_inf_aux(T1, T2),
+  t_is_equal(T1, Inf).
 
 -spec t_is_instance(erl_type(), erl_type()) -> boolean().
 
 %% XXX. To be removed.
```
[the remark this answers is not in the document]
Comment (jhogberg) [reply lib/dialyzer/src/erl_types.erl:3440]: `git blame` says that comment was added by @uabboli 12 years ago, I think it's safe to remove it now :D

File: lib/dialyzer/src/erl_types.erl [hunk 5]
```diff
@@ -2458,36 +2039,39 @@ t_sup1([], Type) ->
...
+         {T1, T2, Res}),
+  Res.
+
+t_sup_aux(?any, _) -> ?any;
+t_sup_aux(_, ?any) -> ?any;
+t_sup_aux(?none, T) -> T;
+t_sup_aux(T, ?none) -> T;
+t_sup_aux(?unit, T) -> T;
+t_sup_aux(T, ?unit) -> T;
+t_sup_aux(T, T) -> do_not_subst_all_vars_to_any(T);
```
[the remark this answers is not in the document]
Comment (jhogberg) [reply lib/dialyzer/src/erl_types.erl:2056]: I suppose not, it's just a small reminder to explore the possibility of handling unification in `t_inf/2` et al. I have some ideas on how to speed up the constraint solving by doing so.

[the remark this answers is not in the document]
Comment (lucioleKi) [reply lib/dialyzer/src/erl_types.erl:2056]: If I understand the comment for this function correctly, it shouldn't be needed anymore. I'll remove it.

**Review by bjorng — DISMISSED** [review]

**Review by bjorng — APPROVED** [review]

**Pull request merged**